### PR TITLE
Add azure_private_dns_zone table. Closes #582

### LINKED ISSUE #582: Add table azure_private_dns_zone
Related to #564

We also needs a table for Private DNS Zones (it's a different namespace)

SDK GO: https://github.com/Azure/azure-sdk-for-go/tree/v68.0.0/services/privatedns/mgmt

### PR DESCRIPTION
**Pull request by pdecat**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
./tint.js azure_private_dns_zone
No env file present for the current environment:  staging
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_private_dns_zone []

PRETEST: tests/azure_private_dns_zone

TEST: tests/azure_private_dns_zone
Running terraform
data.azurerm_client_config.current: Reading...
data.azurerm_client_config.current: Read complete after 0s [id=******=]
data.null_data_source.resource: Reading...
data.null_data_source.resource: Read complete after 0s [id=static]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # azurerm_private_dns_zone.named_test_resource will be created
  + resource "azurerm_private_dns_zone" "named_test_resource" {
      + id                                                    = (known after apply)
      + max_number_of_record_sets                             = (known after apply)
      + max_number_of_virtual_network_links                   = (known after apply)
      + max_number_of_virtual_network_links_with_registration = (known after apply)
      + name                                                  = "turbottest45811.local"
      + number_of_record_sets                                 = (known after apply)
      + resource_group_name                                   = "turbottest45811"
      + tags                                                  = {
          + "name" = "turbottest45811.local"
        }
    }

  # azurerm_resource_group.named_test_resource will be created
  + resource "azurerm_resource_group" "named_test_resource" {
      + id       = (known after apply)
      + location = "eastus"
      + name     = "turbottest45811"
      + tags     = {
          + "name" = "turbottest45811"
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + resource_aka    = (known after apply)
  + resource_id     = (known after apply)
  + resource_name   = "turbottest45811"
  + subscription_id = "******"
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 2s [id=/subscriptions/******/resourceGroups/turbottest45811]
azurerm_private_dns_zone.named_test_resource: Creating...
azurerm_private_dns_zone.named_test_resource: Still creating... [10s elapsed]
azurerm_private_dns_zone.named_test_resource: Still creating... [20s elapsed]
azurerm_private_dns_zone.named_test_resource: Still creating... [30s elapsed]
azurerm_private_dns_zone.named_test_resource: Creation complete after 35s [id=/subscriptions/******/resourceGroups/turbottest45811/providers/Microsoft.Network/privateDnsZones/turbottest45811.local]

Warning: Deprecated

  with data.null_data_source.resource,
  on variables.tf line 29, in data "null_data_source" "resource":
  29: data "null_data_source" "resource" {

The null_data_source was historically used to construct intermediate values
to re-use elsewhere in configuration, the same can now be achieved using
locals

(and one more similar warning elsewhere)

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = "azure:///subscriptions/******/resourceGroups/turbottest45811/providers/Microsoft.Network/privateDnsZones/turbottest45811.local"
resource_id = "/subscriptions/******/resourceGroups/turbottest45811/providers/Microsoft.Network/privateDnsZones/turbottest45811.local"
resource_name = "turbottest45811"
subscription_id = "******"

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/******/resourceGroups/turbottest45811/providers/Microsoft.Network/privateDnsZones/turbottest45811.local",
    "name": "turbottest45811.local",
    "region": "global",
    "type": "Microsoft.Network/privateDnsZones"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/******/resourceGroups/turbottest45811/providers/Microsoft.Network/privateDnsZones/turbottest45811.local",
      "azure:///subscriptions/******/resourcegroups/turbottest45811/providers/microsoft.network/privatednszones/turbottest45811.local"
    ],
    "name": "turbottest45811.local",
    "tags": {
      "name": "turbottest45811.local"
    },
    "title": "turbottest45811.local"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/******/resourceGroups/turbottest45811/providers/Microsoft.Network/privateDnsZones/turbottest45811.local",
    "name": "turbottest45811.local"
  }
]
✔ PASSED

POSTTEST: tests/azure_private_dns_zone

TEARDOWN: tests/azure_private_dns_zone

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select
  name,
  resource_group,
  tags
from
  azure_private_dns_zone;
+-----------------------------------------------------------------------+--------------------------------------------------------------+-------------------------------------------------------+
| name                                                                  | resource_group                                               | tags                                                  |
+-----------------------------------------------------------------------+--------------------------------------------------------------+-------------------------------------------------------+
| ********-****-****-****-************.privatelink.westeurope.azmk8s.io | mc_rg-app-indus-aks-ckd_aks-app-indus-euw-aks-ckd_westeurope | {"autoshutdown":"true","env":"aks-ckd","stack":"app"} |
| private.contoso.xyz                                                   | test-pdecat                                                  | {}                                                    |
+-----------------------------------------------------------------------+--------------------------------------------------------------+-------------------------------------------------------+

> select
  name,
  resource_group
from
  azure_private_dns_zone
where
  number_of_record_sets > 1;
+-----------------------------------------------------------------------+--------------------------------------------------------------+
| name                                                                  | resource_group                                               |
+-----------------------------------------------------------------------+--------------------------------------------------------------+
| ********-****-****-****-************.privatelink.westeurope.azmk8s.io | mc_rg-app-indus-aks-ckd_aks-app-indus-euw-aks-ckd_westeurope |
+-----------------------------------------------------------------------+--------------------------------------------------------------+

> select
  name,
  resource_group
from
  azure_private_dns_zone
where
  number_of_virtual_network_links_with_registration = 0;
+-----------------------------------------------------------------------+--------------------------------------------------------------+
| name                                                                  | resource_group                                               |
+-----------------------------------------------------------------------+--------------------------------------------------------------+
| ********-****-****-****-************.privatelink.westeurope.azmk8s.io | mc_rg-app-indus-aks-ckd_aks-app-indus-euw-aks-ckd_westeurope |
| private.contoso.xyz                                                   | test-pdecat                                                  |
+-----------------------------------------------------------------------+--------------------------------------------------------------+
```
</details>

### CODE REVIEW

File: docs/tables/azure_private_dns_zone.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+```sql
+select
+  name,
+  id,
+from
+  azure_dns_zone;
+```
+
+### List private DNS zones
+**FIXME** zone_type is useless
```
Comment (pdecat) [inline docs/tables/azure_private_dns_zone.md:18]: Find another example query as `zone_type` is useless.

File: docs/tables/azure_private_dns_zone.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  id,
+  tags
+from
+  azure_dns_zone
+where
+  zone_type = 'Private';
+```
+
+### List public DNS zones with delegated name servers
+**FIXME** private do not have name_servers
```
Comment (pdecat) [inline docs/tables/azure_private_dns_zone.md:31]: Find another example query as private zones do not have `name_servers`.

**Review by misraved — COMMENTED** [review]
@pdecat I have left minor review comments. Please let me know if you have questions. Thanks once again for the PR 👍.

File: azure/table_azure_private_dns_zone.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	subscriptionID := session.SubscriptionID
+
+	dnsClient := dns.NewZonesClientWithBaseURI(session.ResourceManagerEndpoint, subscriptionID)
+	dnsClient.Authorizer = session.Authorizer
+
+	op, err := dnsClient.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (pdecat) [reply azure/table_azure_private_dns_zone.go]: Done.

File: azure/table_azure_private_dns_zone.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+func getPrivateDNSZone(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getPrivateDNSZone")
+
+	name := d.EqualsQuals["name"].GetStringValue()
+	resourceGroup := d.EqualsQuals["resource_group"].GetStringValue()
+
+	session, err := GetNewSession(ctx, d, "MANAGEMENT")
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (pdecat) [reply azure/table_azure_private_dns_zone.go]: Done.

File: azure/table_azure_private_dns_zone.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	subscriptionID := session.SubscriptionID
+
+	dnsClient := privatedns.NewPrivateZonesClientWithBaseURI(session.ResourceManagerEndpoint, subscriptionID)
+	dnsClient.Authorizer = session.Authorizer
+
+	result, err := dnsClient.List(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (pdecat) [reply azure/table_azure_private_dns_zone.go]: Done.

File: azure/table_azure_private_dns_zone.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		}),
+	}
+}
+
+//// FETCH FUNCTIONS ////
+
+func listPrivateDNSZones(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	session, err := GetNewSession(ctx, d, "MANAGEMENT")
+	if err != nil {
+		plugin.Logger(ctx).Error("azure_private_dns_zone. listPrivateDNSZones", "client_error", err)
```
[the remark this answers is not in the document]
Comment (pdecat) [reply azure/table_azure_private_dns_zone.go]: Done.

**Comment by pdecat** [comment]
Hi @misraved, thanks for the review!

I've applied your suggested changes, PTAL.

**Review by misraved — APPROVED** [review]
Thanks @pdecat for making the changes so quickly 👍.

**Pull request merged**